### ISSUE TITLE
ls --group-directories-first

### ISSUE THREAD

**Issue by ehawman** [nushell/nushell#7881]
### Related problem

Linux `ls` includes a `--group-directories-first` flag, which puts all directories, regardless of their nature, at the top of the `ls` results then lists remaining objects.

In nushell, `ls | sort-by type` approximates this, but puts symlinks in their own distinct category. When browsing for a dir or file I probably don't care if it's a symlink or not (for locating purposes, at least), and having two places to look is less optimal than grouping like with like.

### Describe the solution you'd like

The addition of a `--group-directories-first` flag, which reads through symlinks to determine if they're ultimately representing a dir or a file, and groups them appropriately.

### Describe alternatives you've considered

`ls | sort-by type` 

### Additional context and details

Obviously lowest priority. Exceedingly small QoL change request.

**Comment by fdncred** [comment]
I like this behavior too but we're pretty resistant to adding command line parameters. There's even one or two that we may remove from `ls`. I wonder if there's something we could add to `sort-by` to make this more composable for other commands to benefit from?

**Comment by ehawman** [comment]
If I'm understanding you correctly, you want to reduce parameters/switches/flags in favor of more specialized and numerous root commands?

**Comment by fdncred** [comment]
Yes, in nushell we try to make command pipelines more composable by having specialized commands that can apply to many commands. It's really a balancing act.

**Comment by webbedspace** [comment]
The issue with `sort-by` moving symlinks would be solved if `sort` could be given a closure that provides custom sorting, similar to JS `.sort()` and the like.

**Comment by fdncred** [comment]
Kind of off-topic/on-topic here but, I'm also wondering if we should redo the `-D` parameter of `ls` to just do this functionality. I really don't get the use case of `-D` the way it is now.

**Comment by WindSoilder** [comment]
> I'm also wondering if we should redo the -D parameter of ls to just do this functionality. I really don't get the use case of -D the way it is now.

Sorry I'm objective on this, `-D` is helpful when I want to show the direction information itself.  For myself, I'm normally using it with `-l`, like `ls -Dl dir` to show more information about that directory

**Comment by fdncred** [comment]
@WindSoilder, understood. I didn't get that it was like the `^ls -ld` in bash. For what it's worth, I'd rather it be `-d` than `-D` and remove the `--du` functionality because we have the `du` command.

I think there's a bit of cleanup needed on `ls`. For example, I'm not sure if we need `ls -f` for full names. You can do `ls | get name | path expand` if you want that. There are a few other parameters like this too.

**Comment by stevenxxiu** [comment]
Here's a workaround that groups directories first and accounts for symlinks:

```nushell
ls | sort-by { $in.name | path expand | path type }
```

The above suggestion can now be done with `sort-by --custom { |a, b| ($a.type == dir) != ($b.type == dir) } | table --icons`, but this doesn't quite work. In one directory of mine, it groups directories then files then directories.

**Comment by kmoschcau** [comment]
@stevenxxiu This doesn't work, because it also returns `true`, when `$a` is not a directory, but `$b` is. The closure should also return `false`, when no reordering is needed, to make the sorting stable, keeping previous ordering.

For me this long form truth-table implementation worked:

```nu
ls | sort-by --custom {|a,b|
  if ($a.type == "dir" and $b.type == "dir") {
    false
  } else if ($a.type == "dir" and $b.type != "dir") {
    true
  } else if ($a.type != "dir" and $b.type == "dir") {
    false
  } else {
    false
  }
}

```

Since only one combination needs to return `true`, we can shorten this to:

```nu
ls | sort-by --custom {|a,b| $a.type == "dir" and $b.type != "dir"}
```

**Comment by Bahex** [comment]
This can be further simplified to just
```nushell
ls | sort-by { $in.type != dir }
```

**Comment by WindSoilder** [comment]
I'm going to close it because it can be achieved by custom command easily
